### PR TITLE
Spring cleaning of the manual's build process

### PR DESCRIPTION
**Pull request by steinuil**
This patch cleans up the build process for the manual and its tools. I took care not to break anything or change the outputs of the build (other than renaming a few intermediate files).

* Unused tools were removed, along with old `.*ignore` files coming from other VCS.
* `make clean` should now delete all generated files.
* `make pregen` and `pregen-etex` now only rebuild files as needed rather than invoking ocamldoc every time. The tools are now also only rebuilt as needed, so calling `make pregen` with no changes from the last invocation takes 0.10s rather than ~3.70s.
* Rebuilding just the html manual with no changes to the stdlib reference now takes about 2.5s rather than 7.7s.

I also tried to improve the code quality in the Makefiles a bit.

The manuals are still fully rebuilt with each invocation even if there were no changes, but fixing that would take a bigger rewrite, and it wouldn't make the builds any faster when a single file has changed anyway. Instead I tried to optimize for building the html manual faster, so that iterative changes to the docs will be somewhat quick if you build with `make html`.

### CODE REVIEW

**Comment by pmetzger** [comment]
Thank you so much for working on this, @steinuil! I haven't had a chance to look at all the patches yet, but it certainly looks like it was a significant amount of effort.

File: manual/Makefile [hunk 1]
```diff
@@ -1,32 +1,37 @@
...
-release:
-	cd manual; ${MAKE} release
-#	cd fpcl; ${MAKE} release
+tests: manual
+	$(MAKE) -C tests all
 
-.PHONY: tools
 tools:
-	cd tools; ${MAKE} clean; ${MAKE} all
+	$(MAKE) -C tools all
```
Comment (mseri) [inline manual/Makefile]: Why did you decide to drop the `make clean` step here, just to reduce further the rebuilds?

**Review by mseri — APPROVED** [review]
Wow this is a nice cleanup, in addition of getting rid of some unnecessary rebuilds, the resulting code is more readable and the syntax more uniform. There are still a lot of loops to go through to follow what is going on but I think it is a good step forward.

If I am not mistaken the only thing missing to have a green CI tick, is update the ChangeLog file. I'm not the best person to suggest which section but I believe `### Compiler distribution build system` could be it

File: manual/Makefile (hunk 1 above)
Comment (steinuil) [reply manual/Makefile]: Yes, I'm not really sure why the tools used to get rebuilt each time because I'm pretty sure they don't depend on anything that changes with each rebuild of the manual.

Comment (gasche) [reply manual/Makefile]: The `make clean` makes sense from the perspective of someone who changes the compiler codebase often, and builds the manual rarely. What's at risk of changing is not the manual build artefacts, but the compiler itself (which is used to build the tools). If the manual/tools Makefile hasn't properly specified its dependencies on the compiler tools (which is hard to do: compilers, standard library, etc.), then it is safer to just rebuild them each time we rebuild the manual, to guarantee that they have been built with the latest compiler version. Otherwise you could get into weird issues when, for example, we change the executable magic numbers, where the tools are compiled with an older compiler version and they refuse to run with the new OCaml runtime during the manual build.

(Then it makes sense to have this clean in this `manual/` Makefile rather than in the `all:` rule in `manual/tools`, as the latter is likely to be used by people who quickly iterate changing the tools, while the former is used by everyone trying to build the manual and not familiar with the subtleties of the build system.)

To summarize: there could be a better way to do this, but it sort of makes sense to me that the `clean` rule is there.

File: manual/tests/Makefile [hunk 2]
```diff
@@ -2,19 +2,26 @@ TOPDIR=$(abspath ../..)
...
-	$(OCAMLC) $(TOPDIR)/compilerlibs/ocamlcommon.cma -I $(TOPDIR)/parsing \
-         -I $(TOPDIR)/driver \
-        cross_reference_checker.ml -o cross-reference-checker
+	$(OCAMLC) $(TOPDIR)/compilerlibs/ocamlcommon.cma \
+	  -I $(TOPDIR)/parsing -I $(TOPDIR)/driver \
+	  $< -o $@
 
+.PHONY: check-cross-references
 check-cross-references: cross-reference-checker
 	$(OCAMLRUN) ./cross-reference-checker \
```
Comment (gasche) [inline manual/tests/Makefile]: While we are at it, this `$(OCAMLRUN)` invocation should probably be `$(SET_LD_PATH) $(OCAMLRUN)` as the others; see 482e8a8b52ff9fb8e8f8aecb17264f314b9a37f0.

**Comment by gasche** [comment]
I wonder about removing dead tools. It is a reasonable idea, but it may be possible that their code would be useful for people interesting in writing new tools. @Octachron, you are the manual-tools expert; do you think the tool codebases removed in this PR had any value?

File: manual/Makefile (hunk 1 above)
Comment (steinuil) [reply manual/Makefile]: Right, I didn't consider that. It shouldn't add much to the build times anyway, since there's not many tools to build anymore.

**Comment by steinuil** [comment]
I asked Octachron a few days ago on the IRC and he said it was probably a good idea, and most of the old tools are just old versions of the current ones from what I could tell.

File: manual/Makefile (hunk 1 above)
Comment (pmetzger) [reply manual/Makefile]: Just an aside: while fixing the manual, I often re-build it many times in an hour, so getting the dependencies right eventually would be best.

**Comment by Octachron** [comment]
My opinion is indeed that the dead tools should be removed or transferred to somewhere else. More precisely, amongst those tools, we have
- caml-tex: a perl precursor of caml_tex2
- format-intf: an ocamldoc precursor
- htmlquote.c: a variant/precursor of texquote2.c that was already not used at the time of the migration from Caml Special Light
- dvi2text, texexpand, latex*, htmlthread, htmltbl, , htmlcut : some hevea  precursor tools in perl or OCaml which are not used since f44aa9930abcf455f .

Consequently, I don't think that those tools would be really useful to anyone trying to write new tools for the OCaml manual.

**Comment by steinuil** [comment]
The last commit should have fixed all those problems. Should I squash some commits together or is it fine like this?

**Comment by pmetzger** [comment]
@steinuil My taste is always to squash commits so the main history of a project remains maximally readable.

**Comment by steinuil** [comment]
@pmetzger that's what I would do, but I made a lot of changes in this PR so I thought that squashing it all together would decrease readability rather than aid it.

Anyway, I squashed together what I thought it'd make sense to have together. If it's all good I'll add an entry to the changelog. Is the `### Compiler distribution build system` section ok?

**Comment by Octachron** [comment]
The build system section of `Changes.md` is the right section.

**Comment by pmetzger** [comment]
Looks like there are conflicts now...

File: manual/tools/Makefile [hunk 3]
```diff
@@ -4,51 +4,35 @@ include $(TOPDIR)/Makefile.tools
 
 CFLAGS=-g -O
 
-all: texquote2 transf htmlquote htmlgen dvi2txt caml-tex2
+all: texquote2 transf caml-tex2
 
-dvi2txt:
-	cd dvi_to_txt; ${MAKE}
 
 transf: transf.cmo htmltransf.cmo transfmain.cmo
```
Comment (Octachron) [inline manual/tools/Makefile:10]: Note for later: htmltransf is not used, it is only used if `transf` is called with  the  `-html` option, which is not the case anymore.

File: manual/manual/Makefile [hunk 4]
```diff
@@ -1,111 +1,97 @@
...
+	  && TEXINPUTS=$(TEXINPUTS) pdflatex pdfmanual.tex
 
-html: files
-	cd htmlman; \
-	mkdir -p libref ; \
+index::
+	cd texstuff \
+	  && sh ../../tools/fix_index.sh pdfmanual.idx \
+	  && makeindex pdfmanual.idx \
+	  && makeindex pdfmanual.kwd.idx
```
Comment (Octachron) [inline manual/manual/Makefile]: Would it be possible to avoid the duplication of  the rules of `index:: … `  ?

File: manual/manual/Makefile [hunk 5]
```diff
@@ -1,111 +1,97 @@
...
+
+html: htmlman/libref/style.css etex-files
+	cd htmlman \
+	  && $(HEVEA) $(HTML_FLAGS) \
+	    -I .. -I ../cmds -I ../library -I ../refman -I ../tutorials \
+	    -I ../../styles -I ../texstuff \
+	    manual.hva -e macros.tex ../manual.tex \
+	  && $(HACHA) -tocter manual.html
+
+htmlman/libref/style.css: style.css $(STDLIB_MLIS) | unprefix_stdlib_for_ocamldoc
```
Comment (Octachron) [inline manual/manual/Makefile:58]: I think it would be nice to mention in a comment  that `htmlman/libref/style.css` is used as witness for the generation of the html doc by ocamldoc.

File: manual/manual/library/Makefile [hunk 6]
```diff
@@ -1,8 +1,28 @@
...
+SRC = ../../..
 
-CSLDIR=$(RELEASEDIR)
+CSLDIR = $(SRC)
 
-STDLIB_INTF= Arg.tex Array.tex ArrayLabels.tex Char.tex Complex.tex \
+LD_PATH := "$(SRC)/otherlibs/unix/:$(SRC)/otherlibs/str/"
+SET_LD_PATH = CAML_LD_LIBRARY_PATH=$(LD_PATH)
+
+VPATH = ".:$(STDLIB_DIR):$(CSLDIR)/parsing:$(CSLDIR)/otherlibs/unix:$(CSLDIR)/otherlibs/str:$(CSLDIR)/otherlibs/graph:$(CSLDIR)/otherlibs/threads:$(CSLDIR)/otherlibs/dynlink:$(CSLDIR)/otherlibs/bigarray"
```
Comment (Octachron) [inline manual/manual/library/Makefile]: Note: this VPATH is no longer used due to the acrobatic way the unprefixed stdlib mlis and cmis are currently produced.

**Review by Octachron — APPROVED** [review]
LGTM: I like the changes, they improve the readability of the Makefiles, and fix few dependencies issues. I only have few minor remarks or self notes on dead code.

**Comment by steinuil** [comment]
Thanks for the review, I made the changes and squashed them into the previous commits.

**Comment by pmetzger** [comment]
@steinuil There are some conflicts, FYI. You need to fix them so that the merge can happen.

**Comment by gasche** [comment]
Merged. Thanks everyone for your work and feedback.

**Comment by xavierleroy** [comment]
> My opinion is indeed that the dead tools should be removed or transferred to somewhere else.

As the author of most of those tools, I agree.  For the most part they were made obsolete by Hevea.
